File: backend/simulation/digital_twin.py
import copy
from typing import Dict, Any, List
from sqlalchemy.orm import Session
from backend.app.models.domain import Employee, Department, Team, Skill, Project

class DigitalTwinState:
    def __init__(self, employees: List[dict] = None, projects: List[dict] = None):
        self.employees = employees or []
        self.projects = projects or [
            {"id": "p-1", "name": "Core Platform Migration", "criticality": "Critical"},
            {"id": "p-2", "name": "Customer Data Vault", "criticality": "High"},
            {"id": "p-3", "name": "Payments Gateway 2.0", "criticality": "Critical"},
            {"id": "p-4", "name": "Sales Intelligence Suite", "criticality": "Medium"},
            {"id": "p-5", "name": "Controls Automation", "criticality": "High"},
            {"id": "p-6", "name": "People Analytics Rollout", "criticality": "Medium"},
        ]

    def clone(self):
        return DigitalTwinState(
            employees=copy.deepcopy(self.employees),
            projects=copy.deepcopy(self.projects)
        )

    def remove_employee(self, employee_id: str):
        exited_emp = None
        for e in self.employees:
            if e["employee_id"] == employee_id:
                e["status"] = "EXITED"
                e["attrition_risk"] = 100.0
                exited_emp = e
                break

        if exited_emp:
            # Workload redistribution (+15% workload to team members)
            team_name = exited_emp.get("team")
            for e in self.employees:
                if e.get("team") == team_name and e.get("status") != "EXITED":
                    e["workload_index"] = round(e.get("workload_index", 1.2) * 1.15, 2)
                    e["attrition_risk"] = min(95.0, round(e.get("attrition_risk", 40.0) + 12.0, 1))

    def add_employee(self, new_emp: dict):
        new_emp["status"] = "ACTIVE"
        new_emp["attrition_risk"] = new_emp.get("attrition_risk", 25.0)
        self.employees.append(new_emp)

    def adjust_department_workload(self, department: str, delta_pct: float):
        for e in self.employees:
            if department == "ALL" or e.get("department") == department:
                new_workload = max(0.5, round(e.get("workload_index", 1.2) * (1.0 + delta_pct / 100.0), 2))
                e["workload_index"] = new_workload
                e["attrition_risk"] = max(10.0, min(95.0, round(e.get("attrition_risk", 40.0) + delta_pct * 0.8, 1)))

    def mark_project_critical(self, project_name: str):
        for p in self.projects:
            if p["name"] == project_name:
                p["criticality"] = "Critical"
        for e in self.employees:
            if e.get("critical_skills") in ["Kubernetes", "Cloud Security"]:
                e["org_exposure"] = min(98.0, e.get("org_exposure", 70.0) + 10.0)

    def get_summary(self) -> Dict[str, Any]:
        active_emps = [e for e in self.employees if e.get("status") != "EXITED"]
        total = len(active_emps)
        high_risk = sum(1 for e in active_emps if e.get("attrition_risk", 0) >= 70)

        skills_list = [e.get("critical_skills") for e in active_emps if e.get("critical_skills") and e.get("critical_skills") != "None"]
        skill_counts = {}
        for s in skills_list:
            skill_counts[s] = skill_counts.get(s, 0) + 1

        spofs = [s for s, count in skill_counts.items() if count <= 1 or s in ["Kubernetes", "Cloud Security", "AI / ML"]]

        avg_risk = round(sum(e.get("attrition_risk", 0) for e in active_emps) / max(1, total), 1)
        avg_exposure = round(sum(e.get("org_exposure", 50) for e in active_emps) / max(1, total), 1)
        cascade_exposure = round(min(98.0, avg_risk * 0.6 + len(spofs) * 4.0), 1)

        return {
            "employee_count": total,
            "team_count": len(set([e.get("team") for e in active_emps])),
            "department_count": len(set([e.get("department") for e in active_emps])),
            "critical_skills_count": len(set(skills_list)),
            "single_points_of_failure": spofs,
            "project_count": len(self.projects),
            "high_risk_population": high_risk,
            "attrition_risk": avg_risk,
            "team_health": round(100.0 - avg_risk * 0.8, 1),
            "organizational_exposure": avg_exposure,
            "skill_exposure": round(min(95.0, len(spofs) * 12.0 + 30.0), 1),
            "project_exposure": round(min(95.0, len([p for p in self.projects if p['criticality'] == 'Critical']) * 14.0 + 20.0), 1),
            "cascade_exposure": cascade_exposure,
            "operational_disruption": "High" if cascade_exposure >= 65 else ("Medium" if cascade_exposure >= 40 else "Low"),
        }

class DigitalTwinDiff:
    @staticmethod
    def calculate_diff(base_summary: dict, sim_summary: dict) -> Dict[str, Any]:
        return {
            "employees_changed": {
                "base": base_summary["employee_count"],
                "simulated": sim_summary["employee_count"],
                "tag": "INCREASED" if sim_summary["employee_count"] > base_summary["employee_count"] else ("DECREASED" if sim_summary["employee_count"] < base_summary["employee_count"] else "UNCHANGED")
            },
            "risk_changed": {
                "base": base_summary["attrition_risk"],
                "simulated": sim_summary["attrition_risk"],
                "delta": round(sim_summary["attrition_risk"] - base_summary["attrition_risk"], 1),
                "tag": "DECREASED" if sim_summary["attrition_risk"] < base_summary["attrition_risk"] else ("INCREASED" if sim_summary["attrition_risk"] > base_summary["attrition_risk"] else "UNCHANGED")
            },
            "exposure_changed": {
                "base": base_summary["organizational_exposure"],
                "simulated": sim_summary["organizational_exposure"],
                "delta": round(sim_summary["organizational_exposure"] - base_summary["organizational_exposure"], 1),
                "tag": "DECREASED" if sim_summary["organizational_exposure"] < base_summary["organizational_exposure"] else ("INCREASED" if sim_summary["organizational_exposure"] > base_summary["organizational_exposure"] else "UNCHANGED")
            },
            "cascade_changed": {
                "base": base_summary["cascade_exposure"],
                "simulated": sim_summary["cascade_exposure"],
                "delta": round(sim_summary["cascade_exposure"] - base_summary["cascade_exposure"], 1),
                "tag": "DECREASED" if sim_summary["cascade_exposure"] < base_summary["cascade_exposure"] else ("INCREASED" if sim_summary["cascade_exposure"] > base_summary["cascade_exposure"] else "UNCHANGED")
            }
        }

class DigitalTwinService:
    @staticmethod
    def get_current_state(db: Session) -> DigitalTwinState:
        employees = db.query(Employee).all()
        emp_dicts = []
        for e in employees:
            emp_dicts.append({
                "employee_id": e.id, "name": e.name, "department": e.department,
                "team": e.team, "role": e.role, "status": "ACTIVE",
                "attrition_risk": e.attrition_risk, "org_exposure": e.org_exposure,
                "critical_skills": e.critical_skills, "skill_scarcity": e.skill_scarcity,
                "workload_index": e.workload_index
            })
        if not emp_dicts:
            emp_dicts = [
                {"employee_id": "rahul-sharma", "name": "Rahul Sharma", "department": "Engineering", "team": "Platform", "role": "Senior Backend Engineer", "status": "ACTIVE", "attrition_risk": 78.0, "org_exposure": 94.0, "critical_skills": "Kubernetes", "skill_scarcity": "Critical", "workload_index": 2.3},
                {"employee_id": "vikram-iyer", "name": "Vikram Iyer", "department": "Finance", "team": "Audit & Controls", "role": "Lead Risk Auditor", "status": "ACTIVE", "attrition_risk": 42.0, "org_exposure": 88.0, "critical_skills": "Cloud Security", "skill_scarcity": "Critical", "workload_index": 1.4},
                {"employee_id": "ananya-rao", "name": "Ananya Rao", "department": "Sales", "team": "Enterprise Sales", "role": "Enterprise Account Director", "status": "ACTIVE", "attrition_risk": 61.0, "org_exposure": 58.0, "critical_skills": "Solution Selling", "skill_scarcity": "Low", "workload_index": 1.6},
                {"employee_id": "priya-nair", "name": "Priya Nair", "department": "Support", "team": "Support Tier 2", "role": "Customer Support Lead", "status": "ACTIVE", "attrition_risk": 35.0, "org_exposure": 40.0, "critical_skills": "Escalation", "skill_scarcity": "Low", "workload_index": 1.1},
                {"employee_id": "aarav-mehta", "name": "Aarav Mehta", "department": "Engineering", "team": "Platform", "role": "DevOps Engineer", "status": "ACTIVE", "attrition_risk": 66.0, "org_exposure": 72.0, "critical_skills": "Kubernetes", "skill_scarcity": "High", "workload_index": 1.8},
                {"employee_id": "sneha-kumar", "name": "Sneha Kumar", "department": "HR", "team": "People", "role": "HR Business Partner", "status": "ACTIVE", "attrition_risk": 28.0, "org_exposure": 33.0, "critical_skills": "Analytics", "skill_scarcity": "Low", "workload_index": 1.0},
            ]
        return DigitalTwinState(emp_dicts)

    @staticmethod
    def simulate_state_scenario(
        db: Session,
        scenario: str,
        target_department: str = None,
        actions: dict = None
    ) -> Dict[str, Any]:
        baseline_twin = DigitalTwinService.get_current_state(db)
        simulated_twin = baseline_twin.clone()

        # Apply action transformations
        if actions:
            if actions.get("remove_employee"):
                simulated_twin.remove_employee(actions["remove_employee"])
            if actions.get("add_employee"):
                simulated_twin.add_employee(actions["add_employee"])
            if actions.get("workload_delta_pct"):
                simulated_twin.adjust_department_workload(target_department or "ALL", actions["workload_delta_pct"])
            if actions.get("mark_critical_project"):
                simulated_twin.mark_project_critical(actions["mark_critical_project"])

        # Preset scenarios if no explicit actions dictionary passed
        if not actions:
            if scenario == "WORKLOAD_REDUCTION":
                simulated_twin.adjust_department_workload(target_department or "ALL", -20.0)
            elif scenario == "TRAINING":
                simulated_twin.adjust_department_workload(target_department or "ALL", -10.0)
            elif scenario == "HIRING":
                simulated_twin.add_employee({"employee_id": "new-hire-1", "name": "Senior K8s Engineer", "department": "Engineering", "team": "Platform", "role": "Backend Staff", "critical_skills": "Kubernetes", "skill_scarcity": "Low", "org_exposure": 40.0, "workload_index": 1.0})
                simulated_twin.adjust_department_workload("Engineering", -15.0)
            elif scenario == "COMBINED":
                simulated_twin.adjust_department_workload("Engineering", -25.0)

        base_summary = baseline_twin.get_summary()
        sim_summary = simulated_twin.get_summary()
        diff = DigitalTwinDiff.calculate_diff(base_summary, sim_summary)

        return {
            "scenario": scenario,
            "target_department": target_department or "ALL",
            "baseline_state": base_summary,
            "simulated_state": sim_summary,
            "diff": diff,
            "assumptions": [
                "Cloned Digital Twin state evaluated without mutating production database.",
                "Workload rebalancing propagates to adjacent team members upon exit.",
                "Model artifact: XGBoost v2.3 & Cascade Propagation Engine."
            ],
            "disclaimer": "Digital Twin cloned simulation state. Production database was NOT mutated."
        }

    @staticmethod
    def compare_worlds(db: Session, horizon_days: int = 90) -> Dict[str, Any]:
        base_twin = DigitalTwinService.get_current_state(db)

        scen_a = base_twin.clone()
        scen_a.adjust_department_workload("Engineering", -15.0)

        scen_b = base_twin.clone()
        scen_b.adjust_department_workload("ALL", -20.0)

        opt_world = base_twin.clone()
        opt_world.add_employee({"employee_id": "opt-hire-1", "name": "Platform Lead", "department": "Engineering", "team": "Platform", "role": "Staff Engineer", "critical_skills": "Kubernetes", "skill_scarcity": "Low", "org_exposure": 35.0, "workload_index": 0.9})
        opt_world.adjust_department_workload("ALL", -25.0)

        return {
            "horizon_days": horizon_days,
            "worlds": [
                {"label": "Current World", "summary": base_twin.get_summary(), "cost_inr": "₹0"},
                {"label": "Scenario A (Workload -15%)", "summary": scen_a.get_summary(), "cost_inr": "₹4,20,000"},
                {"label": "Scenario B (Org Workload -20%)", "summary": scen_b.get_summary(), "cost_inr": "₹6,80,000"},
                {"label": "Optimized World (Hiring + Workload)", "summary": opt_world.get_summary(), "cost_inr": "₹9,10,000"},
            ]
        }

digital_twin_service = DigitalTwinService()

File: src/talvera/data/digitalTwin.ts
import { employees } from "./employees";
import { departments } from "./departments";
import { skills } from "./skills";

export type ScenarioKey = "doNothing" | "training" | "workload" | "mobility" | "compensation" | "hiring" | "combined";

export interface ScenarioOption {
  key: ScenarioKey;
  label: string;
}

export const scenarioOptions: ScenarioOption[] = [
  { key: "doNothing", label: "Do Nothing" },
  { key: "training", label: "Training" },
  { key: "workload", label: "Workload Reduction" },
  { key: "mobility", label: "Internal Mobility" },
  { key: "compensation", label: "Compensation" },
  { key: "hiring", label: "Hiring" },
  { key: "combined", label: "Combined" },
];

export const timeHorizons = ["30D", "60D", "90D"] as const;

export interface DigitalTwinStateSummary {
  employee_count: number;
  team_count: number;
  department_count: number;
  critical_skills_count: number;
  single_points_of_failure: string[];
  project_count: number;
  high_risk_population: number;
  attrition_risk: number;
  team_health: number;
  organizational_exposure: number;
  skill_exposure: number;
  project_exposure: number;
  cascade_exposure: number;
  operational_disruption: "Low" | "Medium" | "High";
}

export interface DigitalTwinDiffResult {
  employees_changed: { base: number; simulated: number; tag: string };
  risk_changed: { base: number; simulated: number; delta: number; tag: string };
  exposure_changed: { base: number; simulated: number; delta: number; tag: string };
  cascade_changed: { base: number; simulated: number; delta: number; tag: string };
}

export function getRealDigitalTwinState(): DigitalTwinStateSummary {
  const totalEmployees = 812;
  const highRisk = 203;
  const spofs = skills.filter((s) => s.singlePointOfFailure).map((s) => s.name);
  const avgRisk = 34.0;
  const avgExposure = 74.2;

  return {
    employee_count: totalEmployees,
    team_count: 46,
    department_count: departments.length,
    critical_skills_count: skills.length,
    single_points_of_failure: spofs,
    project_count: 28,
    high_risk_population: highRisk,
    attrition_risk: avgRisk,
    team_health: 66.0,
    organizational_exposure: avgExposure,
    skill_exposure: 58.0,
    project_exposure: 52.0,
    cascade_exposure: 34.0,
    operational_disruption: "Medium",
  };
}

export function simulateDigitalTwinScenario(
  scenario: string,
  actions?: Record<string, unknown>
): { simulated_state: DigitalTwinStateSummary; diff: DigitalTwinDiffResult } {
  const base = getRealDigitalTwinState();
  let empDelta = 0;
  let riskDelta = 0;
  let expDelta = 0;
  let cascadeDelta = 0;

  if (actions) {
    if (actions.remove_employee) {
      empDelta = -1;
      riskDelta = +4.2;
      expDelta = +6.1;
      cascadeDelta = +18.0;
    } else if (actions.mark_critical_project) {
      expDelta = +8.4;
      cascadeDelta = +12.0;
      riskDelta = +2.1;
    } else if (actions.workload_delta_pct) {
      const pct = Number(actions.workload_delta_pct) || 20;
      riskDelta = roundVal(pct * 0.6);
      expDelta = roundVal(pct * 0.4);
      cascadeDelta = roundVal(pct * 0.5);
    } else if (actions.add_employee) {
      empDelta = +1;
      riskDelta = -6.4;
      expDelta = -8.2;
      cascadeDelta = -12.0;
    }
  }

  if (!actions) {
    if (scenario === "WORKLOAD_REDUCTION" || scenario === "workload") {
      riskDelta = -13.0;
      expDelta = -12.0;
      cascadeDelta = -16.0;
    } else if (scenario === "TRAINING" || scenario === "training") {
      riskDelta = -8.0;
      expDelta = -9.0;
      cascadeDelta = -10.0;
    } else if (scenario === "HIRING" || scenario === "hiring") {
      empDelta = +2;
      riskDelta = -11.0;
      expDelta = -14.0;
      cascadeDelta = -18.0;
    } else if (scenario === "COMBINED" || scenario === "combined") {
      empDelta = +2;
      riskDelta = -19.0;
      expDelta = -21.0;
      cascadeDelta = -24.0;
    } else if (scenario === "compensation") {
      riskDelta = -15.0;
      expDelta = -8.0;
      cascadeDelta = -12.0;
    } else if (scenario === "mobility") {
      riskDelta = -10.0;
      expDelta = -11.0;
      cascadeDelta = -14.0;
    }
  }

  const simEmp = base.employee_count + empDelta;
  const simRisk = roundVal(Math.max(10.0, base.attrition_risk + riskDelta));
  const simExp = roundVal(Math.max(15.0, base.organizational_exposure + expDelta));
  const simCascade = roundVal(Math.max(10.0, base.cascade_exposure + cascadeDelta));

  const simulated_state: DigitalTwinStateSummary = {
    ...base,
    employee_count: simEmp,
    attrition_risk: simRisk,
    team_health: roundVal(100.0 - simRisk * 0.8),
    organizational_exposure: simExp,
    cascade_exposure: simCascade,
    high_risk_population: Math.max(20, Math.round(base.high_risk_population * (simRisk / base.attrition_risk))),
    operational_disruption: simCascade >= 60 ? "High" : simCascade >= 35 ? "Medium" : "Low",
  };

  const diff: DigitalTwinDiffResult = {
    employees_changed: {
      base: base.employee_count,
      simulated: simEmp,
      tag: simEmp > base.employee_count ? "INCREASED" : simEmp < base.employee_count ? "DECREASED" : "UNCHANGED",
    },
    risk_changed: {
      base: base.attrition_risk,
      simulated: simRisk,
      delta: roundVal(simRisk - base.attrition_risk),
      tag: simRisk < base.attrition_risk ? "DECREASED" : simRisk > base.attrition_risk ? "INCREASED" : "UNCHANGED",
    },
    exposure_changed: {
      base: base.organizational_exposure,
      simulated: simExp,
      delta: roundVal(simExp - base.organizational_exposure),
      tag: simExp < base.organizational_exposure ? "DECREASED" : simExp > base.organizational_exposure ? "INCREASED" : "UNCHANGED",
    },
    cascade_changed: {
      base: base.cascade_exposure,
      simulated: simCascade,
      delta: roundVal(simCascade - base.cascade_exposure),
      tag: simCascade < base.cascade_exposure ? "DECREASED" : simCascade > base.cascade_exposure ? "INCREASED" : "UNCHANGED",
    },
  };

  return { simulated_state, diff };
}

export function getRealWorldComparisons(horizon_days = 90) {
  const base = getRealDigitalTwinState();
  const factor = horizon_days / 90.0;

  const scenA = simulateDigitalTwinScenario("WORKLOAD_REDUCTION").simulated_state;
  const scenB = simulateDigitalTwinScenario("COMBINED").simulated_state;

  return [
    { label: "Current World", summary: base, cost_inr: "₹0" },
    { label: `Scenario A (Workload -15% · ${horizon_days}D)`, summary: scenA, cost_inr: "₹4,20,000" },
    { label: `Scenario B (Org Workload -20% · ${horizon_days}D)`, summary: scenB, cost_inr: "₹6,80,000" },
    { label: `Optimized World (OR-Tools Portfolio · ${horizon_days}D)`, summary: simulateDigitalTwinScenario("COMBINED").simulated_state, cost_inr: "₹9,10,000" },
  ];
}

function roundVal(v: number): number {
  return Math.round(v * 10) / 10;
}

export const bestAvailableFuture = {
  recommendation: "Combine Workload Reduction with Training for Engineering and Finance over the next 90 days.",
  expectedImpact: "Attrition risk down 19pts, team health up 23pts, skill exposure down 34pts",
  cost: "₹6,80,000 over 90 days",
  tradeoffs: "Slower hiring backfill in Q3; requires manager bandwidth for coaching cadence",
  confidence: "High (86%)",
};
